runtime/engine/src/test/java/io/aklivity/zilla/runtime/engine/test/internal/catalog/TestCatalogHandler.java

```diff
@@ -95,4 +97,63 @@ public String resolve(
     {
         return schema;
     }
+
+    @Override
+    public int decode(
+        DirectBuffer data,
+        int index,
+        int length,
+        ValueConsumer next,
+        SchemaConfig catalog,
+        String subject,
+        Read read)
+    {
+        int schemaId;
+        int progress = 0;
+        int valLength = -1;
+        if (data.getByte(index) == MAGIC_BYTE)
+        {
+            progress += BitUtil.SIZE_OF_BYTE;
+            schemaId = data.getInt(index + progress, ByteOrder.BIG_ENDIAN);
+            progress += BitUtil.SIZE_OF_INT;
+        }
+        else if (catalog != null && catalog.id != NO_SCHEMA_ID)
+        {
+            schemaId = catalog.id;
+        }
+        else
+        {
+            schemaId = resolve(subject, catalog.version);
+        }
+
+        if (schemaId > NO_SCHEMA_ID)
+        {
+            valLength = read.accept(data, index + progress, length - progress, next, schemaId);
+        }
+        return valLength;
+    }
+
+    @Override
+    public int resolve(
+        DirectBuffer data,
+        int index,
+        int length,
+        SchemaConfig catalog,
+        String subject)
+    {
+        int schemaId;
+        if (data.getByte(index) == MAGIC_BYTE)
+        {
+            schemaId = data.getInt(index + BitUtil.SIZE_OF_BYTE, ByteOrder.BIG_ENDIAN);
+        }
+        else if (catalog.id != NO_SCHEMA_ID)
+        {
+            schemaId = catalog.id;
+        }
+        else
+        {
+            schemaId = resolve(subject, catalog.version);
+        }
+        return schemaId;
+    }
```

Comment:
Let's see if we can simplify this to not require embedding magic byte and schema id in payload.